src/main/java/org/example/cosmocats/service/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+package org.example.cosmocats.service.impl;
+
+import org.example.cosmocats.repository.entity.CategoryEntity;
+import org.example.cosmocats.common.CategoryType;
+import org.example.cosmocats.repository.CategoryRepository;
+import org.example.cosmocats.service.CategoryService;
+import org.example.cosmocats.service.exception.CategoryNotFoundException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+@Transactional
+public class CategoryServiceImpl implements CategoryService {
+
+    private final CategoryRepository categoryRepository;
+
+    public CategoryServiceImpl(CategoryRepository categoryRepository) {
+        this.categoryRepository = categoryRepository;
+    }
+
+    @Override
+    public CategoryEntity createCategory(CategoryEntity category) {
+
+        if (categoryRepository.findByType(category.getType()).isPresent()) {
+            throw new RuntimeException("Category with type " + category.getType() + " already exists");
+        }
+
+        return categoryRepository.save(category);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<CategoryEntity> getAllCategories() {
+        return categoryRepository.findAll();
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public Optional<CategoryEntity> getCategoryById(Long id) {
+        return categoryRepository.findById(id);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public Optional<CategoryEntity> getCategoryByType(CategoryType type) {
+        return categoryRepository.findByType(type);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<CategoryEntity> searchCategoriesByDescription(String keyword) {
+        return categoryRepository.findByDescriptionContaining(keyword);
+    }
+
+    @Override
+    public CategoryEntity updateCategory(Long categoryId, CategoryEntity category) {
```

Comment:
problem here, you don't consider general exceptions that could potentially raise. Like db connectivity issue. So what you should do: try {} catch (Exception ex ) // that will hadle all exceptions coming from the category, than you should throw wrapper exception like PersistanceException. so later you can handle it accordingly and send to client 500 - DB connectivity lost or whatever. Also in case when you are throwing exceptions like here - you would add extra catch (CategoryNotFoundException ex) throw ex - so you will still throw the exception